File: android/src/main/java/com/github/wuxudong/rncharts/data/PieDataExtract.java
package com.github.wuxudong.rncharts.data;

import android.graphics.Color;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableType;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;
import com.github.mikephil.charting.interfaces.datasets.IDataSet;
import com.github.wuxudong.rncharts.utils.BridgeUtils;
import com.github.wuxudong.rncharts.utils.ChartDataSetConfigUtils;
import com.github.wuxudong.rncharts.utils.ConversionUtil;

import java.util.ArrayList;

/**
 * Created by xudong on 02/03/2017.
 */

public class PieDataExtract extends DataExtract<PieData, PieEntry> {
    @Override
    PieData createData() {
        return new PieData();
    }

    @Override
    IDataSet<PieEntry> createDataSet(ArrayList<PieEntry> entries, String label) {
        return new PieDataSet(entries, label);
    }

    @Override
    void dataSetConfig(IDataSet<PieEntry> dataSet, ReadableMap config) {
        PieDataSet pieDataSet = (PieDataSet) dataSet;
        ChartDataSetConfigUtils.commonConfig(pieDataSet, config);

        if (BridgeUtils.validate(config, ReadableType.Boolean, "xValuePosition")) {
            Boolean isDrawOutSide = config.getBoolean("xValuePosition");
            if (isDrawOutSide == true) {
                pieDataSet.setXValuePosition(PieDataSet.ValuePosition.OUTSIDE_SLICE);
            } else {
                pieDataSet.setXValuePosition(PieDataSet.ValuePosition.INSIDE_SLICE);
            }
        }
        if (BridgeUtils.validate(config, ReadableType.String, "valueLineColor")) {
            String hex = config.getString("valueLineColor");
            pieDataSet.setValueLineColor(Color.parseColor(hex));
        }
        if (BridgeUtils.validate(config, ReadableType.Number, "valueLineWidth")) {
            pieDataSet.setValueLineWidth((float) config.getDouble("valueLineWidth"));
        }
        if (BridgeUtils.validate(config, ReadableType.Number, "valueLinePart1Length")) {
            pieDataSet.setValueLinePart1Length((float) config.getDouble("valueLinePart1Length"));
        }
        if (BridgeUtils.validate(config, ReadableType.Number, "valueLinePart2Length")) {
            pieDataSet.setValueLinePart2Length((float) config.getDouble("valueLinePart2Length"));
        }
        // PieDataSet only config
        if (BridgeUtils.validate(config, ReadableType.Number, "sliceSpace")) {
            pieDataSet.setSliceSpace((float) config.getDouble("sliceSpace"));
        }
        if (BridgeUtils.validate(config, ReadableType.Number, "selectionShift")) {
            pieDataSet.setSelectionShift((float) config.getDouble("selectionShift"));
        }
    }

    @Override
    PieEntry createEntry(ReadableArray values, int index) {
        PieEntry entry;

        if (ReadableType.Map.equals(values.getType(index))) {
            ReadableMap map = values.getMap(index);

            float value = (float) map.getDouble("value");
            if (BridgeUtils.validate(map, ReadableType.String, "label")) {
                entry = new PieEntry(value, map.getString("label"), ConversionUtil.toMap(map));
            } else {
                entry = new PieEntry(value, ConversionUtil.toMap(map));
            }
        } else if (ReadableType.Number.equals(values.getType(index))) {
            entry = new PieEntry((float) values.getDouble(index));
        } else {
            throw new IllegalArgumentException("Unexpected entry type: " + values.getType(index));
        }

        return entry;
    }

}

File: ios/ReactNativeCharts/pie/PieDataExtract.swift
//  Created by xudong wu on 02/03/2017.
//  Copyright © 2017 wuxudong. All rights reserved.
//

import Foundation

import SwiftyJSON
import Charts
import UIKit

extension UIColor {
  convenience init?(hexString: String) {
    var chars = Array(hexString.hasPrefix("#") ? hexString.characters.dropFirst() : hexString.characters)  // Swift 4 just use `var chars = Array(hexString.hasPrefix("#") ? hexString.dropFirst() : hexString[...])`
    let red, green, blue, alpha: CGFloat
    switch chars.count {
    case 3:
      chars = chars.flatMap { [$0, $0] }
      fallthrough
    case 6:
      chars = ["F","F"] + chars
      fallthrough
    case 8:
      alpha = CGFloat(strtoul(String(chars[0...1]), nil, 16)) / 255
      red   = CGFloat(strtoul(String(chars[2...3]), nil, 16)) / 255
      green = CGFloat(strtoul(String(chars[4...5]), nil, 16)) / 255
      blue  = CGFloat(strtoul(String(chars[6...7]), nil, 16)) / 255
    default:
      return nil
    }
    self.init(red: red, green: green, blue:  blue, alpha: alpha)
  }
}

class PieDataExtract : DataExtract {
    override func createData() -> ChartData {
        return PieChartData();
    }
    
    override func createDataSet(_ entries: [ChartDataEntry]?, label: String?) -> IChartDataSet {
        return PieChartDataSet(values: entries, label: label)
    }
    
    override func dataSetConfig(_ dataSet: IChartDataSet, config: JSON) {
        let pieDataSet = dataSet as! PieChartDataSet;
      
        if (config["xValuePosition"].bool != nil) {
          var isDrawOutSide : Bool
          isDrawOutSide = (config["xValuePosition"].bool)!
          if (isDrawOutSide == true) {
            pieDataSet.xValuePosition = .outsideSlice
          } else {
            pieDataSet.xValuePosition = .insideSlice
          }
        }
      
        if config["valueLineWidth"].float != nil {
            let valueConfig = CGFloat(config["valueLineWidth"].float!)
            pieDataSet.valueLineWidth = valueConfig
        }
      
        if config["valueLinePart1Length"].float != nil {
            let valueConfig = CGFloat(config["valueLinePart1Length"].float!)
            pieDataSet.valueLinePart1Length = valueConfig
        }
      
        if config["valueLinePart2Length"].float != nil {
            let valueConfig = CGFloat(config["valueLinePart2Length"].float!)
            pieDataSet.valueLinePart1Length = valueConfig
        }
      
        if config["valueLineColor"].string != nil {
            let valueConfig = config["valueLineColor"].string
            pieDataSet.valueLineColor = UIColor(hexString: valueConfig!)
        }
      
        ChartDataSetConfigUtils.commonConfig(pieDataSet, config: config);
        
        // PieDataSet only config
        if config["sliceSpace"].number != nil {
            pieDataSet.sliceSpace = CGFloat(config["sliceSpace"].numberValue)
        }
        
        if config["selectionShift"].number != nil {
            pieDataSet.selectionShift = CGFloat(config["selectionShift"].numberValue)
        }
  }
    
    override func createEntry(_ values: [JSON], index: Int) -> ChartDataEntry {
        var entry: PieChartDataEntry;
        
        let item = values[index];
        
        if item.dictionary != nil {
            let dict = item;
            
            var value : Double;
            if dict["value"].double != nil {
                value = Double((dict["value"].doubleValue));
            } else {
                fatalError("invalid data " + values.description);
            }
            
            if dict["label"].string != nil {
                entry = PieChartDataEntry(value: value, label: dict["label"].stringValue)
            } else {
                entry = PieChartDataEntry(value: value)
            }
            
            entry.data = dict as AnyObject?
        } else if item.double != nil {
            entry = PieChartDataEntry(value : item.doubleValue);
        } else {
            fatalError("invalid data " + values.description);
        }
        
        return entry;
    }
}
